src/itemdef.cpp

```diff
@@ -319,103 +318,72 @@ class CItemDefManager: public IWritableItemDefManager
 	}
 #ifndef SERVER
 public:
-	ClientCached* createClientCachedDirect(const std::string &name,
-			Client *client) const
+	ClientCached* createClientCachedDirect(const ItemStack &item, Client *client) const
 	{
-		infostream<<"Lazily creating item texture and mesh for \""
-				<<name<<"\""<<std::endl;
+		IItemDefManager *idef = client->getItemDefManager();
+		std::string inventory_image = item.getInventoryImage(idef);
+		std::string item_name = item.getDefinition(idef).name;
+		std::string cache_key = item_name;
+		if (!inventory_image.empty())
+			cache_key += "/" + inventory_image;
+
+		infostream << "Lazily creating item texture and mesh for \""
+				<< cache_key << "\""<<std::endl;
 
 		// This is not thread-safe
 		sanity_check(std::this_thread::get_id() == m_main_thread);
 
 		// Skip if already in cache
-		ClientCached *cc = NULL;
-		m_clientcached.get(name, &cc);
-		if(cc)
-			return cc;
+		auto it = m_clientcached.find(cache_key);
+		if (it != m_clientcached.end())
+			return it->second.get();
 
 		ITextureSource *tsrc = client->getTextureSource();
-		const ItemDefinition &def = get(name);
 
 		// Create new ClientCached
-		cc = new ClientCached();
+		auto cc = std::make_unique<ClientCached>();
 
 		// Create an inventory texture
 		cc->inventory_texture = NULL;
-		if (!def.inventory_image.empty())
-			cc->inventory_texture = tsrc->getTexture(def.inventory_image);
-
-		ItemStack item = ItemStack();
-		item.name = def.name;
-
-		getItemMesh(client, item, &(cc->wield_mesh));
+		if (inventory_image.empty())
+			getItemMesh(client, item, &(cc->wield_mesh));
+		else
+			cc->inventory_texture = tsrc->getTexture(inventory_image);
 
+		const ItemDefinition &def = get(item_name);
```

Comment:
Could be moved up (`item.getDefinition(idef)`) (and replace this line: `std::string item_name = item.getDefinition(idef).name;`).